amqp_broker.go

```diff
@@ -10,98 +10,90 @@ import (
-		Durable:    true,
-		AutoDelete: false,
-	}
-}
-
 //AMQPCeleryBroker is Broker client for AMQP
 type AMQPCeleryBroker struct {
-	connection *amqp.Connection
+	connection       *amqp.Connection
+	exchange         *AMQPExchange
+	defaultSendQueue *AMQPQueue
+	prefetchCount    int
+	listener         chan interface{}
+	listenerQueues   []string
+	connect          func() (bool, error)
+}
 
-	exchange      *AMQPExchange
-	queue         *AMQPQueue
-	prefetchCount int
-	listener      <-chan amqp.Delivery
-	connect       func() error
+type AMQPBrokerConfig struct {
+	//URL url of amqp broker
+	URL string
+	//Exchange optional non-default exchange name
+	Exchange string
+	//DefaultSendQueue optional - all send calls will by default be routed to this queue
+	DefaultSendQueue string
+	ConnectionConfig *amqp.Config
 }
 
 // NewAMQPCeleryBroker creates new AMQPCeleryBroker using AMQP conn and channel
-func NewAMQPCeleryBroker(host string, config *amqp.Config, queueName string) (*AMQPCeleryBroker, error) {
+func NewAMQPCeleryBroker(config *AMQPBrokerConfig) (*AMQPCeleryBroker, error) {
+	if err := validateConfig(config); err != nil {
+		return nil, err
+	}
 	broker := &AMQPCeleryBroker{
-		exchange:      NewAMQPExchange(exchangeName),
-		queue:         NewAMQPQueue(queueName),
-		prefetchCount: 4,
+		exchange:         NewAMQPExchange(config.Exchange),
+		defaultSendQueue: NewAMQPQueue(config.DefaultSendQueue),
+		listener:         make(chan interface{}),
+		prefetchCount:    4,
 	}
 	connMutex := new(sync.Mutex)
-	broker.connect = func() error {
+	broker.connect = func() (bool, error) {
 		connMutex.Lock()
 		defer connMutex.Unlock()
-		return broker.lazyConnect(host, config)
+		return broker.lazyConnect(config.URL, config.ConnectionConfig)
 	}
 	return broker, nil
 }
 
 // Listen spawns receiving channel on AMQP queue
-func (b *AMQPCeleryBroker) Listen() error {
-	channel, err := b.openChannel()
-	if err != nil {
-		return err
+func (b *AMQPCeleryBroker) Listen(queues ...string) error {
+	b.listenerQueues = append(b.listenerQueues, queues...)
+	if len(b.listenerQueues) == 0 {
+		return nil
 	}
-	b.listener, err = channel.Consume(b.queue.Name, "", false, false, false, true, nil)
-	if err != nil {
+	if reconnected, err := b.connect(); err != nil {
 		return err
+	} else if !reconnected {
+		return b.registerConsumers(queues)
 	}
 	return nil
 }
 
 // SendCeleryMessage sends CeleryMessage to broker
-func (b *AMQPCeleryBroker) SendCeleryMessage(message *CeleryMessage, _ ...string) error {
+func (b *AMQPCeleryBroker) SendCeleryMessage(message *CeleryMessage) error {
+	if _, err := b.connect(); err != nil {
+		return err
+	}
+	taskMessage := message.GetTaskMessage()
 	channel, err := b.openChannel()
 	if err != nil {
 		return err
 	}
 	defer func() {
 		err = channel.Close()
 		if err != nil {
-			fmt.Println("error closing channel")
+			logrus.Warnf("error closing channel")
 		}
 	}()
-	taskMessage := message.GetTaskMessage()
+	queue := b.defaultSendQueue
```

Comment:
If we are always sending to the defaultSendQueue, but we can [listen on several queues](https://github.com/pnongah/gocelery/blob/b2b718941bf426e47d3db60e28655eef2f1142bf/amqp_broker.go#L61) based on the task;  how do we get a message to a particular task on a particular queue?

Reply:
we don't always send on the default. It's adjustable at the Delay() API call level. Look at the fields on the gocelery.TaskParameters struct: it has a Queue field which the consumer can adjust on a per-function call basis. If not specified, we fall back to this one.